Ad9545 add model

tachicialex opened the pull request:
Add chip model for AD9545.

3f298a3a0f41 - pyadi-jiff: examples: Add AD9545
a0c9b3923e68 - pyadi-jiff: add model for AD9545
409488aa5ef8 - pyaadi-jiff: Add AD9545 support

tfcollins commented on adijif/clocks/ad9545.py:
```diff
@@ -0,0 +1,330 @@
+        self._add_equation(equations)
+
+        """ Instruct solver to try to avoid Minimum and Maximum PLL rates """
+        for i in range(0, 2):
```
Please add a flag to disable optimizations/cost functions. For larger problems, these need to get converted to lexicographic versions

tfcollins commented on adijif/clocks/ad9545.py:
```diff
@@ -0,0 +1,330 @@
+        Returns:
+            bool: Always False
+        """
+        self.model.options.SOLVER = 1  # APOPT solver
```
Do that default settings from clock.py not work for gekko?

tachicialex replied:
With the default settings from clock.py , for some inputs i get solution not found.

tfcollins replied:
If you update clocks.py to these settings do other tests fail?

tachicialex replied:
So i added this in clock.py: 
        self.model.options.SOLVER = 1  # APOPT solver
        self.model.solver_options = [
            "minlp_maximum_iterations 300000000",
            "minlp_max_iter_with_int_sol 6000",
            "minlp_as_nlp 0",
            "minlp_branch_method 1",
            "minlp_integer_tol 0.0001"
        ]
And when running 'make test' everything seems to be fine.

tfcollins commented on examples/ad9545_example.py:
```diff
@@ -0,0 +1,23 @@
+
+input_refs = [1, 10e6, 0, 0]
+
+output_clocks = [30720000, 0, 0 , 0, 0, 0, 0 ,0 ,0, 0]
```
Do you have to put zeros? or can you just have a smaller list?

tachicialex replied:
The zeros are necessary or i could change the method to accept an array of tuples (input_number, rate).

tfcollins replied:
I would recommend accepting only clocks the user cares about aka not zero pad. This is how the other clock chips function. But also limit the user to 10 output clocks

tfcollins commented:
Please add a test for the AD9545. Something like this: https://github.com/analogdevicesinc/pyadi-jif/blob/main/tests/test_clocks.py#L10

tachicialex commented:
Changelog V1 - > V2:
- added flags to disable cost functions
- removed clk_names arg.
- fixed multiple CPLEX objectives bug
- changed GEKKO solver_options after trying multiple usecases
- added tests for ad9545

tfcollins commented on tests/test_clocks.py:
```diff
@@ -6,6 +6,59 @@
+
+    clk.set_requested_clocks(input_refs, output_clocks)
+
+    clk.solve()
```
Can you check the config against a reference output here

tachicialex commented:
Changelog V2 -> V3:
- changed zero padded inputs to tuples: (ref_nr, rate_hz)
- when adding integer_var changed domain to normal min, max. For some reason
CPLEX would always return infeasible when using domains.
- added solution checks in test_ad9545_validate_pass

tachicialex commented:
Changelog V3 -> V4:
- instead of raw solution comparison, check if rate formulas are ok
CPLEX will give different answers on different hardware
